### PR TITLE
group_by(<computation>) always done on the ungrouped data

### LINKED ISSUE #5938: Original grouping ignored in group_by(...)
Hi there

I have encountered an unexpected behaviour with `group_by(...)`. Expressions in `...` ignore the original grouping in an already grouped data frame.

Example: group an already grouped data frame further into two equally populated bins calculated separately for each group. 

Below snippet does not lead to the expected result. The calculation of bins ignores the initial grouping, bins are calculated over the entire data frame instead:
```r
library(tidyverse)

iris %>%
    group_by(.data$Species) %>%
    group_by(bin = cut(.data$Petal.Width, 2), .add = TRUE) %>% 
    summarize(n = n(), .groups = "drop")
## # A tibble: 4 x 3
##   Species    bin              n
##   <fct>      <fct>        <int>
## 1 setosa     (0.0976,1.3]    50
## 2 versicolor (0.0976,1.3]    28
## 3 versicolor (1.3,2.5]       22
## 4 virginica  (1.3,2.5]       50
```

Below snippet works as expected. Bins are calculated independently for each group:
```r
iris %>%
    group_by(.data$Species) %>%
    mutate(bin = cut(.data$Petal.Width, 2)) %>% 
    group_by(.data$bin, .add = TRUE) %>% 
    summarize(n = n(), .groups = "drop")
## # A tibble: 6 x 3
##   Species    bin               n
##   <fct>      <fct>         <int>
## 1 setosa     (0.0995,0.35]    41
## 2 setosa     (0.35,0.6]        9
## 3 versicolor (0.999,1.4]      35
## 4 versicolor (1.4,1.8]        15
## 5 virginica  (1.4,1.95]       21
## 6 virginica  (1.95,2.5]       29
```

This example contradicts a statement on the `group_by` help page that grouping by expressions "is just short-hand for
a mutate() followed by a group_by()".

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #5938

### CODE REVIEW

**Comment by romainfrancois** [comment]
``` r
library(tidyverse)

iris %>%
  group_by(.data$Species) %>%
  group_by(bin = cut(.data$Petal.Width, 2), .add = TRUE) %>% 
  summarize(n = n(), .groups = "drop")
#> # A tibble: 6 × 3
#>   Species    bin               n
#>   <fct>      <fct>         <int>
#> 1 setosa     (0.0995,0.35]    41
#> 2 setosa     (0.35,0.6]        9
#> 3 versicolor (0.999,1.4]      35
#> 4 versicolor (1.4,1.8]        15
#> 5 virginica  (1.4,1.95]       21
#> 6 virginica  (1.95,2.5]       29
```

<sup>Created on 2021-09-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

File: man/nest_by.Rd [hunk 1]
```diff
@@ -11,7 +11,11 @@ nest_by(.data, ..., .key = "data", .keep = FALSE)
 lazy data frame (e.g. from dbplyr or dtplyr). See \emph{Methods}, below, for
 more details.}
 
-\item{...}{In \code{group_by()}, variables or computations to group by.
+\item{...}{In \code{group_by()}, variables or computations to group by,
```
Comment (romainfrancois) [inline man/nest_by.Rd:14]: It is a bit unfortunate that the doc for `...` is imported here, but `...` are forbidden in `nest_by(<grouped_df>)`.

**Comment by romainfrancois** [comment]
I believe the errors are unrelated, have tried to fix in https://github.com/tidyverse/dplyr/pull/6025

**Comment by romainfrancois** [comment]
Apart from the unrelated failures, I think it's better that `group_by(bin = cut(Petal.Width, 2), .add = TRUE)` evaluates `cut(Petal.Width, 2)` within each group ?

**Comment by romainfrancois** [comment]
Yeah it does feel a little complicated. Perhaps we should go as we initially hinted in https://github.com/tidyverse/dplyr/issues/5938#issuecomment-871152987 , always do the `mutate()` on the ungrouped data, and then prepend the original groups if `.add = TRUE`, but document that this is different from calling `mutate()` before `group_by()`.

**Comment by DavisVaughan** [comment]
I would probably lean towards keeping it simpler (i.e. the current CRAN behavior), especially since mutating within a group-by is somewhat uncommon.

Improved documentation would be great, emphasizing that the expressions are always computed on the ungrouped data frame, and that the user should use `mutate()` beforehand if they don't want that. I think that is the easiest thing to explain.

**Comment by romainfrancois** [comment]
I'm still not sure about this, but I tried to add some tests and documentation for the behavior that we have. 

``` r
library(dplyr, warn.conflicts = FALSE)

iris %>%
  group_by(Species) %>%
  group_by(bin = cut(Petal.Width, 2), .add = TRUE) %>% 
  summarize(n = n(), .groups = "drop")
#> # A tibble: 4 × 3
#>   Species    bin              n
#>   <fct>      <fct>        <int>
#> 1 setosa     (0.0976,1.3]    50
#> 2 versicolor (0.0976,1.3]    28
#> 3 versicolor (1.3,2.5]       22
#> 4 virginica  (1.3,2.5]       50

iris %>%
  group_by(Species) %>%
  mutate(bin = cut(Petal.Width, 2)) %>% 
  group_by(bin, .add = TRUE) %>% 
  summarize(n = n(), .groups = "drop")
#> # A tibble: 6 × 3
#>   Species    bin               n
#>   <fct>      <fct>         <int>
#> 1 setosa     (0.0995,0.35]    41
#> 2 setosa     (0.35,0.6]        9
#> 3 versicolor (0.999,1.4]      35
#> 4 versicolor (1.4,1.8]        15
#> 5 virginica  (1.4,1.95]       21
#> 6 virginica  (1.95,2.5]       29
```

<sup>Created on 2021-09-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

For context: looking back at why we decided to do the implicit `mutate()` on the ungrouped data: https://github.com/tidyverse/dplyr/issues/5473 https://github.com/tidyverse/dplyr/pull/5598

**Review by hadley — APPROVED** [review]
LGTM

**Comment by romainfrancois** [comment]
Thanks. Merging as the remaining problems are about `last_warnings()`... so unrelated.

**Pull request merged**